[Bug]: seedless onboarding regression: "Invalid refresh token" when attempting to log in

### Describe the bug

This occured when I was trying to reproduce a test failure we see in extension, where we lock the wallet before some background requests are completed.

I logged in, and then clicked the Lock button as quickly as I could. I did this a few times until this message appeared.

I was trying to reproduce this on a production build (13.0.0 from the chrome web store): https://github.com/MetaMask/metamask-extension/pull/34888#pullrequestreview-3097879446

Restarting Chrome does not fix it.


### Expected behavior

I should be able to log in.

### Screenshots/Recordings

https://github.com/user-attachments/assets/8111b76e-c9ee-424a-89c1-7decc5fc1ca7

### Steps to reproduce

See the attached video demo. I slowed down my network connection in the video; I don't know if it is a required step to repro.

### Error messages or log output

```shell
sentry-install.js:1 error while revoking seedless refresh token (Error#1)
sentry-install.js:1 Error#1: SeedlessOnboardingController - Invalid refresh token

sentry-install.js:1   at oe.B (chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:31239)
  at oe.P (chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:28911)
  at chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:21590
  at async le (chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:22524)
  at async oe._ (chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:29796)
  at async oe.revokeRefreshToken (chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/common-3.js:9:21068)
```

### Detection stage

In production (default)

### Version

13.0.0

### Build type

None

### Browser

Chrome

### Operating system

Windows

### Hardware wallet

_No response_

### Additional context

_No response_

### Severity

_No response_

--- a/app/scripts/services/oauth/base-login-handler.ts
+++ b/app/scripts/services/oauth/base-login-handler.ts
@@ -141,6 +141,10 @@ export abstract class BaseLoginHandler {
       },
     );
 
+    if (!res.ok) {
+      throw new Error('Failed to revoke refresh token');
+    }
+
     const data = await res.json();
     return {
       // TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31860
@@ -171,6 +175,10 @@ export abstract class BaseLoginHandler {
         body: requestData,
       },
     );
+
+    if (!res.ok) {
+      throw new Error('Failed to get auth token');
+    }
 
     const data = await res.json();
     return data;

--- a/app/scripts/services/oauth/oauth-service.test.ts
+++ b/app/scripts/services/oauth/oauth-service.test.ts
@@ -68,6 +68,8 @@ describe('OAuthService - startOAuthLogin', () => {
             // eslint-disable-next-line @typescript-eslint/naming-convention
             id_token: MOCK_JWT_TOKEN,
           }),
+          status: 200,
+          ok: true,
         });
       }) as jest.Mock,
     );
@@ -157,8 +159,7 @@ describe('OAuthService - startOAuthLogin', () => {
 });
 
 describe('OAuthService - getNewRefreshToken', () => {
-  beforeEach(() => {
-    // mock the fetch call to auth-server
+  it('should be able to get new refresh token', async () => {
     jest.spyOn(global, 'fetch').mockImplementation(
       jest.fn(() => {
         return Promise.resolve({
@@ -173,12 +174,12 @@ describe('OAuthService - getNewRefreshToken', () => {
             // eslint-disable-next-line @typescript-eslint/naming-convention
             revoke_token: 'MOCK_NEW_REVOKE_TOKEN',
           }),
+          status: 200,
+          ok: true,
         });
       }) as jest.Mock,
     );
-  });
 
-  it('should be able to get new refresh token', async () => {
     const oauthConfig = loadOAuthConfig();
 
     const oauthService = new OAuthService({
@@ -222,10 +223,34 @@ describe('OAuthService - getNewRefreshToken', () => {
       },
     );
   });
+
+  it('should throw an error if the get refresh token api call fails', async () => {
+    jest.spyOn(global, 'fetch').mockImplementation(
+      jest.fn(() => {
+        return Promise.resolve({
+          status: 401,
+        });
+      }) as jest.Mock,
+    );
+
+    const oauthService = new OAuthService({
+      env: getOAuthLoginEnvs(),
+      webAuthenticator: mockWebAuthenticator,
+      bufferedTrace: mockBufferedTrace,
+      bufferedEndTrace: mockBufferedEndTrace,
+    });
+
+    await expect(
+      oauthService.getNewRefreshToken({
+        connection: AuthConnection.Google,
+        refreshToken: 'MOCK_REFRESH_TOKEN',
+      }),
+    ).rejects.toThrow('Failed to get auth token');
+  });
 });
 
 describe('OAuthService - revokeAndGetNewRefreshToken', () => {
-  beforeEach(() => {
+  it('should be able to get new refresh token', async () => {
     // mock the fetch call to auth-server
     jest.spyOn(global, 'fetch').mockImplementation(
       jest.fn(() => {
@@ -240,12 +265,12 @@ describe('OAuthService - revokeAndGetNewRefreshToken', () => {
             // eslint-disable-next-line @typescript-eslint/naming-convention
             revoke_token: 'MOCK_NEW_REVOKE_TOKEN',
           }),
+          status: 201,
+          ok: true,
         });
       }) as jest.Mock,
     );
-  });
 
-  it('should be able to get new refresh token', async () => {
     const oauthService = new OAuthService({
       env: getOAuthLoginEnvs(),
       webAuthenticator: mockWebAuthenticator,
@@ -278,5 +303,29 @@ describe('OAuthService - revokeAndGetNewRefreshToken', () => {
         }),
       },
     );
+  });
+
+  it('should throw an error if the revoke refresh token api call fails', async () => {
+    jest.spyOn(global, 'fetch').mockImplementation(
+      jest.fn(() => {
+        return Promise.resolve({
+          status: 401,
+        });
+      }) as jest.Mock,
+    );
+
+    const oauthService = new OAuthService({
+      env: getOAuthLoginEnvs(),
+      webAuthenticator: mockWebAuthenticator,
+      bufferedTrace: mockBufferedTrace,
+      bufferedEndTrace: mockBufferedEndTrace,
+    });
+
+    await expect(
+      oauthService.revokeAndGetNewRefreshToken({
+        connection: AuthConnection.Google,
+        revokeToken: 'MOCK_REVOKE_TOKEN',
+      }),
+    ).rejects.toThrow('Failed to revoke refresh token');
   });
 });

--- a/package.json
+++ b/package.json
@@ -334,7 +334,7 @@
     "@metamask/rpc-errors": "^7.0.0",
     "@metamask/safe-event-emitter": "^3.1.1",
     "@metamask/scure-bip39": "^2.0.3",
-    "@metamask/seedless-onboarding-controller": "^2.5.0",
+    "@metamask/seedless-onboarding-controller": "npm:@metamask-previews/seedless-onboarding-controller@2.5.0-preview-bfa447b",
     "@metamask/selected-network-controller": "^22.1.0",
     "@metamask/signature-controller": "^32.0.0",
     "@metamask/smart-transactions-controller": "^17.0.0",

--- a/package.json
+++ b/package.json
@@ -334,7 +334,7 @@
     "@metamask/rpc-errors": "^7.0.0",
     "@metamask/safe-event-emitter": "^3.1.1",
     "@metamask/scure-bip39": "^2.0.3",
-    "@metamask/seedless-onboarding-controller": "npm:@metamask-previews/seedless-onboarding-controller@2.5.0-preview-bfa447b",
+    "@metamask/seedless-onboarding-controller": "^2.5.1",
     "@metamask/selected-network-controller": "^22.1.0",
     "@metamask/signature-controller": "^32.0.0",
     "@metamask/smart-transactions-controller": "^17.0.0",

